app/src/main/java/protect/card_locker/Utils.java

```diff
@@ -208,16 +208,23 @@ static private BarcodeValues getBarcodeFromBitmapReal(Bitmap bitmap) {
         }
     }
 
+    static public Boolean isNotYetValid(Date validFromDate) {
+        return validFromDate.after(getStartOfToday().getTime());
```

Comment:
It's currently working the way you expect it to. The `validFromDate` is stored at midnight in DatePickerFragment's `onDateSet` method (October 16 00:00:00.000) and `getStartOfToday` also returns the today's date at midnight, so on October 16 `after` will return false, i.e. the date is valid.

I realize I can merge the two methods `isNotYetValid` and `hasExpired` into a single `isValid` method, so the code is easier to maintain. What do you think?

Reply:
> It's currently working the way you expect it to. The validFromDate is stored at midnight in DatePickerFragment's onDateSet method (October 16 00:00:00.000) and getStartOfToday also returns the today's date at midnight, so on October 16 after will return false, i.e. the date is valid.

That's not what's happening for me on your branch. If I right now on October 17th create a card that's Valid From October 17th it is marked as red for me:

![image](https://user-images.githubusercontent.com/1885159/196243489-aec8cb93-f13a-4ee6-8153-210c00a4438b.png)

> I realize I can merge the two methods isNotYetValid and hasExpired into a single isValid method, so the code is easier to maintain. What do you think?

I think I'd prefer a name like `isTodayOrLater`, it's a bit clearer than `isValid`.

Reply:
Sorry, you're right. I'll try to fix it asap.

Reply:
I couldn't make it work with a single method like I imagined it 😞 At least I managed to fix the bug.

The `Calendar` instance in `Utils.getStartOfToday` is instantiated as `new GregorianCalendar()` and the one in `DatePickerFragment.onDateSet` is instantiated with `Calendar.getInstance()`. It turns out the default value for `Calendar.AM_PM` is different between them (could be because of the phone's regional settings) and the first one sets the hour at midnight (12:00 AM) while the other sets it at noon (12:00 PM). 

Theoretically, adding something like `c.set(Calendar.AM_PM, Calendar.AM);` to the DatePickerFragment's Calendar could have worked but only for new cards, as the existing cards are already saved in the DB at noon. I'm not too keen on trying not to break the DB with a migration script 😄 Instead, I added a check if the date is not today.

Reply:
Oh geez, that's a subtle bug, nice catch!

I wonder though, with how you're explaining this, wouldn't that mean that the expiryCheck is also broken (firing 12 hours too early or too late)?

I'm kinda worried now about what else may be wrong, I wonder what other calendar types exist and what other subtle bugs might've creeped in with the `Calendar.getInstance()` and `new GregorianCalendar()` difference :fearful:

Reply:
That's great research!

So, if I understand things correctly some small fixes there would be:
1. Maybe change the current `Calendar.getInstance()` to `new GregorianCalendar()` just to be safe and prevent possible bugs due to possible changes in the future (because right now it's not synced up and I don't have any clue how bad breakage may be if Android adds a new calendar in the future and we start comparing between calendar types).
2. Change the code in `DatePickerFragment.onDateSet()` to use `HOUR_OF_DAY` instead of `HOUR` so new database entries are written correctly.
3. Add some note in the `hasExpired()` explaining the previous bug and how it could be 12 hours late for cards saved before the `HOUR_OF_DAY` fix.

It won't fix the weirdness for existing cards, but it'll gradually transition every single user out of this weird little bug I believe.

I would say these fixes would fit in a new PR as they're not directly part of this new feature?

Reply:
Yes, I believe the steps you described will fix the bug for the newly created cards. I'll create a PR.